Merge pin-3.7 into master

samjijina opened the pull request:
This merge will enable MacSim to use PIN 3.13 by default. Also, the x86 traces have been updated along with a new utility called "trace converter" which allows users to convert older PIN traces such as PIN 3.7 to PIN 3.13.

Trace generator has also been upgraded.

hyesoon committed:
converting trace generator to work with pin 3.7

hyesoon committed:
adding the output of xed output from the pin tool

hyesoon committed:
adding new enumerators for pin 3.7 tools

hyesoon committed:
update enumerator generation part

Commit:
added --newpin option

hyesoon committed:
fix library issue (please copy the gzlib  into
    pin-3.7/extras/xed-intel/lib)

fix opcode enumerator names

hyesoon committed:
Merge branch 'pin-3.7' of https://github.com/gthparch/macsim into pin-3.7

hyesoon committed:
instruction mixiture statistics are updated for new pin

Commit:
added selective use of *.stat.def (refer scripts/statgen.pl for more detail)

hyesoon committed:
some cleanning up

hyesoon committed:
x86 new version trace file list

hyesoon committed:
memory hiearchy testing trace generator

hyesoon committed:
adding annotation example for trace generation

hyesoon committed:
update trace generator and macsim enumartors for pin 3.13

hyesoon committed:
adding necessary chages pin 3.13

hyesoon committed:
inst stats for pin 3.13

hyesoon committed:
make trace reader reads traces & update README in the trace generator

hyesoon committed:
adding pin_3_7_trace generation flag set internally main.cc

hyesoon committed:
initial frame to convert traces

hyesoon committed:
add pin 212 to pin 311 opcode enumerator conversion pass

samjijina committed:
fixed readme for trace converter tool (#19)

* Commented stdout code to cleanup terminal

* Fixed README for trace converter

samjijina committed:
Fixed --newpin for build.py and compile errors

samjijina committed:
Fixed 64bit debug printing and other typos

hyesoon committed:
fix initialization array size to compile on newer gcc

samjijina committed:
Fixed typos

samjijina committed:
Fixed clean option to remove .out files from bin dir

samjijina committed:
Removed cleaning of .out files

samjijina committed:
Created new x86 sample trace which will be packaged with all future MacSim versions. This trace was generated using PIN 3.13 (#22)

samjijina committed:
Set default pin and fix (#23)

* Set Pin 3.13 to default in macsim. Fixed some small errors in the enumerators of Pin 3.13

* Fixed README --> added more info

samjijina committed:
fixed example usage of xalancbmk (#25)

* Created new x86 sample trace which will be packaged with all future MacSim versions. This trace was generated using PIN 3.13

* Fixed example usage of xalancbmk and added back xalancbmk as a converted trace

samjijina commented:
Reverted changes using rebase and drop

hyojongk commented on SConscript:
```diff
@@ -73,6 +73,11 @@ if flags['qsim'] == '1':
   env['CPPPATH']    += [os.environ['XED_HOME'] + "/include"]
   env['LIBPATH']    += [os.environ['QSIM_PREFIX'] + "/lib", os.environ['XED_HOME'] + "/lib"]
 
+#if flags['pin_3_7_trace'] == '1':
```
do we still need these comments?

samjijina replied:
resolved

hyojongk commented on bin/params.in:
```diff
@@ -1,10 +1,10 @@
-core_type ptx
-large_core_type igpu
+num_sim_large_cores 1
+core_type x86 
```
we don't use small core for x86

samjijina replied:
@hyesoon Please review this

samjijina replied:
resolved

hyojongk commented on bin/trace_file_list:
```diff
@@ -1,2 +1,3 @@
 1
-../sst-unit-test/traces/cachesize_1/trace.txt
+/home/common/intel_cpu_pin/pin-3.7-97619-g0d0c92f4f-gcc-linux/source/tools/x86_trace_generator/trace.txt
```
these new traces are not part of this repo. use the one in sst-unit-test

samjijina replied:
resolved

hyojongk commented on SConstruct:
```diff
@@ -101,7 +103,7 @@ if flags['ramulator'] == '1':
 
 ## Create stat/knobs
-SConscript('scripts/SConscript')
+SConscript('scripts/SConscript', exports='flags')
```
why do we need this?

hyojongk commented on build.py:
```diff
@@ -23,6 +23,7 @@ def parse_arg():
   parser.add_option("-c", "--clean", action="store_true", dest="clean", default=False, help="clean")
   parser.add_option("-t", "--test", action="store_true", dest="test", default=False, help="clean")
   parser.add_option("-v", "--val", action="store_true", dest="val", default=False, help="build version used for gpu validation")
+  parser.add_option("--newpin", action="store_true", dest="pin_3_7_trace", default=False, help="trace based on pin 3.7 ")
```
change 3.7 to appropriate version which i think is 3.13?

samjijina replied:
resolved

hyojongk commented on def/inst.stat.def:
```diff
@@ -362,5 +402,7 @@ DEF_STAT (OP_CAT_GED_DIM                          , COUNT, NO_RATIO, PER_CORE)
 DEF_STAT (OP_CAT_GED_INVALID                      , DIST,  NO_RATIO, PER_CORE)
 
 
+
```
remove these unnecessary empty lines

samjijina replied:
done

hyojongk commented on macsim.config:
```diff
@@ -3,6 +3,7 @@
 [Build]
 debug: 0
 gprof: 0
+pin_3_7_trace: 0
```
again change 3.7 to 3.13 if we really want to use specific version number.

hyojongk commented on src/trace_read.cc:
```diff
@@ -284,7 +284,7 @@ bool trace_read_c::read_trace(int core_id, void *trace_info, int sim_thread_id,
       }
 
-      if (*KNOB(KNOB_DEBUG_TRACE_READ))
+			if (*KNOB(KNOB_DEBUG_TRACE_READ) || *KNOB(KNOB_DEBUG_PRINT_TRACE)) 
```
check indentation

hyojongk commented on src/trace_read.cc:
```diff
@@ -426,239 +426,602 @@ void trace_read_c::convert_t_uop_to_info(trace_uop_s *t_uop, inst_info_s *info)
-  "rdf",
+   "*invalid*",
+   "*none*",
+   "*UNKNOWN REG 2*",
```
double check if this is correct one

hyojongk commented on src/trace_read.cc:
```diff
@@ -426,239 +426,602 @@ void trace_read_c::convert_t_uop_to_info(trace_uop_s *t_uop, inst_info_s *info)
+   "inst_g29d",
 };
 
+/*
```
delete these lines.

hyojongk commented on src/trace_read.cc:
```diff
@@ -684,8 +1052,84 @@ const char* trace_read_c::g_tr_opcode_names[MAX_TR_OPCODE_NAME] = {
+  "CPU_OPCODE_LAST",
 };
 
+
```
delete these lines

hyojongk commented on src/trace_read.h:
```diff
@@ -53,7 +53,7 @@ POSSIBILITY OF SUCH DAMAGE.
 
 
-#define MAX_TR_REG 300 // REG_LAST = 298
+#define MAX_TR_REG 321 // REG_LAST = 298 // need to check!!! 
```
need to check? delete the comment after confirmation

samjijina replied:
@hyesoon Please review this

hyesoon replied:
@samjijina can you count and double check?

samjijina replied:
done resolved

hyojongk commented on src/trace_read.h:
```diff
@@ -275,68 +275,111 @@ typedef struct trace_uop_s {
 ///////////////////////////////////////////////////////////////////////////////////////////////
 typedef enum TR_TEMP_TEMP_ENUM_ {
-  TR_REG_TMP0 = 167
+  TR_REG_TMP0 =256  
```
add whitespace

hyojongk commented on src/trace_read.h:
```diff
@@ -275,68 +275,111 @@ typedef struct trace_uop_s {
 /// \brief Enumerator ID for temp register
 ///
-/// Currently, we have 166 registers, so temp register will be 167.
+/// Currently, Pin 3.7 has 330 registers but we only store up to 255 register numbers. 
```
3.13

hyojongk commented on src/trace_read.h:
```diff
@@ -356,6 +399,91 @@ typedef enum CPU_OPCODE_ENUM_ {
 } CPU_OPCODE_ENUM;
 
 
+/*
```
delete this

hyojongk commented on src/trace_read_cpu.cc:
```diff
@@ -148,40 +148,82 @@ bool cpu_decoder_c::ungetch_trace(int core_id, int sim_thread_id, int num_inst)
 void cpu_decoder_c::dprint_inst(void *trace_info, int core_id, int thread_id) 
 {
-  if (m_dprint_count++ >= 50000 || !*KNOB(KNOB_DEBUG_PRINT_TRACE))
+	  if (m_dprint_count++ >= 50000 || !*KNOB(KNOB_DEBUG_PRINT_TRACE))
```
check indentation

hyojongk commented on src/trace_read_cpu.cc:
```diff
@@ -1206,373 +1248,961 @@ bool cpu_decoder_c::get_uops_from_traces(int core_id, uop_c *uop, int sim_thread
+  m_int_uop_table[XED_CATEGORY_BROADCAST] = UOP_IADD;
+  m_int_uop_table[XED_CATEGORY_CALL] = UOP_IADD;
+  m_int_uop_table[XED_CATEGORY_CET] = UOP_IADD; // new
+	m_int_uop_table[XED_CATEGORY_CLDEMOTE] = UOP_IADD; // new
```
check indentation

hyojongk commented on src/trace_read_cpu.cc:
```diff
@@ -1206,373 +1248,961 @@ bool cpu_decoder_c::get_uops_from_traces(int core_id, uop_c *uop, int sim_thread
 }
 
 
+/*
```
delete

hyojongk commented on src/trace_read_cpu.cc:
```diff
@@ -1206,373 +1248,961 @@ bool cpu_decoder_c::get_uops_from_traces(int core_id, uop_c *uop, int sim_thread
-  "st7",
-  "r_status_flags",
-  "rdf",
+	"*invalid*", /* 0*/ 
```
check indentation

hyojongk commented on src/trace_read_cpu.cc:
```diff
@@ -1206,373 +1248,961 @@ bool cpu_decoder_c::get_uops_from_traces(int core_id, uop_c *uop, int sim_thread
+   "x87",
 };
 
+/*
```
delete this

hyojongk commented on src/trace_read_cpu.cc:
```diff
@@ -1598,8 +2233,114 @@ const char* cpu_decoder_c::g_tr_opcode_names[MAX_TR_OPCODE_NAME] = {
+  "CPU_OPCODE_LAST",
 };
 
+/*
```
delete this

hyojongk commented on src/trace_read_cpu.cc:
```diff
@@ -1614,7 +2355,7 @@ const char* cpu_decoder_c::g_tr_cf_names[10] = {
 };
 
-
+// divide instruvction should be included and separated 
```
instruction

hyojongk commented:
xalancbmk is also x86

hyojongk commented:
tools/trace_converter/all_knobs cc and h are supposed to be generated by the script. so those files should not be added.

hyojongk commented on tools/trace_reader/main.cc:
```diff
@@ -49,6 +49,8 @@ do {                                                              \
   }                                                               \
 } while (0)
 
+#define PIN_3_7_TRACE 
```
3_13?

hyojongk commented on tools/trace_reader/main.cc:
```diff
@@ -188,9 +190,9 @@ int main(int argc, char* argv[])
   KnobsContainer* knob_container = new KnobsContainer();
   g_knobs = knob_container->getAllKnobs();
-  knob_container->applyParamFile("params.in");
+  // knob_container->applyParamFile("params.in");
```
why comment this and below out?

samjijina replied:
@hyojongk Because @hyesoon fixed trace_reader by not reading the params.in file. According to Trello, it needs to be fixed later as a bug report

hyesoon replied:
yes for now, just comment out for reading params.in file. now that is causing a seg fault. to be fixed in future.

samjijina replied:
okay am resolving this then

hyojongk commented on tools/x86_trace_generator/memtest.cp:
```diff
@@ -0,0 +1,218 @@
+#include <cstdlib>
```
file extension correct?

hyojongk commented on tools/x86_trace_generator/pin3.7_enumerator.txt:
```diff
@@ -0,0 +1,422 @@
+typedef enum CPU_OPCODE_ENUM_ {
```
3.13

hyojongk commented on tools/x86_trace_generator/trace_generator.cpp:
```diff
@@ -32,18 +32,18 @@ POSSIBILITY OF SUCH DAMAGE.
 
-#include "../../src/hmc_types.h"
+#include "pin.H"
+//#include "../InstLib/time_warp.H"
```
if we don't need these any more just delete them.

hyojongk commented on tools/x86_trace_generator/trace_generator.cpp:
```diff
@@ -52,9 +52,9 @@ POSSIBILITY OF SUCH DAMAGE.
-#if defined(TARGET_IA32) || defined(TARGET_IA32E)
-TIME_WARP tw;
-#endif
+//#if defined(TARGET_IA32) || defined(TARGET_IA32E)
```
if we don't need these lines any more delete them

hyojongk commented on tools/x86_trace_generator/trace_generator.cpp:
```diff
@@ -141,7 +140,8 @@ UINT32 thread_count = 0;
 unsigned int g_inst_print_count[MAX_THREADS];
 unsigned int func_count = 0;
-CONTROL control;
+// CONTROL control;
```
if we don't need this, delete it

hyojongk commented on tools/x86_trace_generator/trace_generator.cpp:
```diff
@@ -688,7 +664,11 @@ void instrument(INS ins)
         info->is_fp = TRUE;
 
-      *ptr = REG_FullRegName(*begin);
+			uint16_t tmp = static_cast<uint16_t>(REG_FullRegName(*begin));
```
check indentation

hyojongk commented on tools/x86_trace_generator/trace_generator.cpp:
```diff
@@ -715,7 +695,12 @@ void instrument(INS ins)
         info->is_fp = 1;
 
-      *ptr = REG_FullRegName(*begin);
+
```
check indentation

hyojongk commented on tools/x86_trace_generator/trace_generator.cpp:
```diff
@@ -798,34 +783,39 @@ void instrument(INS ins)
-  else {
-    info->cf_type = NOT_CF;
-  }
+  // if (INS_IsIndirectBranchOrCall(ins) && !INS_IsRet(ins) && !INS_IsInterrupt(ins)) {
```
we use two whitespace for indentation not four. check indentation

hyojongk commented on tools/x86_trace_generator/trace_generator.h:
```diff
@@ -69,15 +69,15 @@ struct Thread_info
 };
 
-
+// new version april-2019
```
remove unnecessary comments

hyojongk commented on tools/x86_trace_generator/trace_generator.h:
```diff
@@ -69,15 +69,15 @@ struct Thread_info
   uint8_t  num_dest_regs;       // 3-bits
-  uint8_t  src[MAX_SRC_NUM];    // 6-bits * 4
-  uint8_t  dst[MAX_DST_NUM];    // 6-bits * 4
+  uint8_t  src[MAX_SRC_NUM];    // incresed in 2019 version // 6-bits * 4 // back to 8 
```
increased

hyojongk commented on tools/x86_trace_generator/trace_generator.h:
```diff
@@ -161,59 +160,97 @@ string tr_cf_names[15] = {
-string tr_opcode_names[66] = {
-  "INVALID",
+string tr_opcode_names[106] = {
+	  "INVALID",
```
indentation

hyojongk commented on tools/x86_trace_generator/xed_extractor.cpp:
```diff
@@ -212,12 +221,12 @@ void macsim_pin_convert(void)
-//  extract_opcode(); // useless
-//  macsim_opcode();
-//  macsim_opcode_enum();
+ // extract_category();
```
check indentation and delete unnecessary lines

hyojongk reviewed (changes requested):
please address my comments

samjijina committed:
Resolved SConscript comments

samjijina committed:
Resolved trace_file_list comments

samjijina committed:
Resolved build.py comments

samjijina committed:
Resolved inst.stat.def comments

samjijina committed:
Resolved macsim.config

samjijina committed:
Resolved trace_read.cc

samjijina committed:
Resolved other issues

samjijina committed:
fixed memtest file extension

samjijina committed:
renamed the enumerator file

samjijina committed:
resolved trace generator comments

samjijina committed:
resolved trace generator comments

samjijina committed:
resolved trace generator comments part 2

samjijina committed:
fixed xed extractor comments

samjijina committed:
fixing build issues

samjijina committed:
fixed script

samjijina committed:
verified MAX_TR_REG value

samjijina committed:
fixed PERL warnings

hyojongk commented on src/trace_read_cpu.cc:
```diff
@@ -152,36 +152,78 @@ void cpu_decoder_c::dprint_inst(void *trace_info, int core_id, int thread_id)
+  *m_dprint_output << "*** end of the data structure *** " << endl << endl;
+	
+	/*
+   // screen output debugging 
```
do we need to keep this?

samjijina replied:
@hyojongk Yes please

hyojongk commented on src/trace_read.cc:
```diff
@@ -71,88 +69,81 @@ POSSIBILITY OF SUCH DAMAGE.
 
 #ifdef USING_QSIM
   using namespace Qsim;
-  if (!m_tg) {
```
i think this is against our format guideline

samjijina replied:
@hyojongk What exactly seems to be the problem? Is it the use of the ! operator? or the naming convention? 
Thanks

hyojongk commented on src/trace_read_cpu.cc:
```diff
@@ -1614,7 +1916,7 @@ const char* cpu_decoder_c::g_tr_cf_names[10] = {
 };
 
-
+// divide instruction should be included and separated 
```
meaning?

samjijina replied:
resolved

hyojongk reviewed (dismissed):
1. please drop changes to the internal submodule from this PR.
2. bin/params.in -- we don't use small core for x86.
3. trace_read.cc (and probably other files) -- we have format guideline (doc/macsim.pdf secsion A). please follow it.
4. drop tools/trace_converter/all_knobs.cc and h files from this PR. those files are supposed to be generated by knobgen.pl using knob.def.

samjijina committed:
untracked the all_knobs.cc and header file

samjijina committed:
fixed the core type for x86 trace

samjijina commented:
> 1. please drop changes to the internal submodule from this PR.
> 2. bin/params.in -- we don't use small core for x86.
> 3. trace_read.cc (and probably other files) -- we have format guideline (doc/macsim.pdf secsion A). please follow it.
> 4. drop tools/trace_converter/all_knobs.cc and h files from this PR. those files are supposed to be generated by knobgen.pl using knob.def.

2. fixed
4. fixed
3. need followup
1. I need your help for that, it seems like the submodule is updating with the commits you made in the past

ramyadhadidi committed:
Merge branch 'master' into pin-3.7

ramyadhadidi reviewed (approved):
Done in the meeting!
